pkg/liqo-controller-manager/external-network/wireguard/wggatewayserver_controller.go

```diff
@@ -0,0 +1,369 @@
+
+	networkingv1alpha1 "github.com/liqotech/liqo/apis/networking/v1alpha1"
+	"github.com/liqotech/liqo/pkg/consts"
+	"github.com/liqotech/liqo/pkg/discovery"
+	"github.com/liqotech/liqo/pkg/utils"
+)
+
+// WgGatewayServerReconciler manage WgGatewayServer lifecycle.
+type WgGatewayServerReconciler struct {
+	client.Client
+	Scheme           *runtime.Scheme
+	extNetPodsClient client.Client
+}
+
+// NewWgGatewayServerReconciler returns a new WgGatewayServerReconciler.
+func NewWgGatewayServerReconciler(cl client.Client, s *runtime.Scheme, extNetPodsClient client.Client) *WgGatewayServerReconciler {
+	return &WgGatewayServerReconciler{
+		Client:           cl,
+		Scheme:           s,
+		extNetPodsClient: extNetPodsClient,
+	}
+}
+
+// +kubebuilder:rbac:groups=networking.liqo.io,resources=wggatewayservers,verbs=get;list;watch;delete;create;update;patch
+// +kubebuilder:rbac:groups=networking.liqo.io,resources=wggatewayservers/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;delete;create;update;patch
+// +kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;delete;create;update;patch
+// +kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch
+
+// Reconcile manage GatewayServer lifecycle.
+func (r *WgGatewayServerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, err error) {
+	wgServer := &networkingv1alpha1.WgGatewayServer{}
+	if err = r.Get(ctx, req.NamespacedName, wgServer); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.V(4).Infof("WireGuard gateway server %q not found", req.NamespacedName)
+			return ctrl.Result{}, nil
+		}
+		klog.Errorf("Unable to get the WireGuard gateway server %q: %v", req.NamespacedName, err)
+		return ctrl.Result{}, err
+	}
+
+	requeue := false
+	depNsName := types.NamespacedName{Namespace: wgServer.Namespace, Name: wgServer.Name}
+	svcNsName := types.NamespacedName{Namespace: wgServer.Namespace, Name: wgServer.Name}
+
+	if wgServer.DeletionTimestamp.IsZero() {
+		needStatusUpdate := false
+
+		// Ensure deployment (create or update)
+		dep, err := r.ensureDeployment(ctx, wgServer, depNsName)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		// Ensure service (create or update)
+		_, err = r.ensureService(ctx, wgServer, svcNsName)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		// Handle status
+		if err := r.handleEndpointStatus(ctx, wgServer, svcNsName, dep, &needStatusUpdate, &requeue); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if err := r.handleSecretRefStatus(ctx, wgServer, &needStatusUpdate, &requeue); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		// update status
+		if needStatusUpdate {
+			if err := r.Status().Update(ctx, wgServer); err != nil {
+				klog.Errorf("Unable to update status of %q: %v", req.NamespacedName, err)
+				return ctrl.Result{}, err
+			}
+			klog.Infof("%q status correctly updated", req.NamespacedName)
```

Comment:
The order should be:
```
defer updateStatus
handle
handle
```

I think that we can update the status every time, it is not a problem... if we want to avoid unnecessary updates we can perform a deepequal check here instead of doing it in the other functions, but i don't think that it is necessary